smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java

```diff
@@ -0,0 +1,375 @@
+    }
+
+    /**
+     * Request a new upload slot with optional content type from custom upload service.
+     *
+     * When you get slot you should upload file to PUT URL and share GET URL.
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @param filename name of file to be uploaded
+     * @param fileSize file size in bytes -- must be less or equal
+     *                 to {@link HttpFileUploadManager#getMaxFileSize()} (if available)
+     * @param contentType file content-type or null
+     * @param uploadService upload service to use or null for default one
+     * @return file upload Slot in case of success
+     * @throws IllegalArgumentException if fileSize is less than or equal to zero
+     *                                  or greater than {@link HttpFileUploadManager#getMaxFileSize()}
+     * @throws SmackException
+     * @throws InterruptedException
+     * @throws XMPPException.XMPPErrorException
+     */
+    public Slot requestSlot(String filename, long fileSize, String contentType, DomainBareJid uploadService)
+            throws SmackException, InterruptedException, XMPPException.XMPPErrorException {
+        if (defaultUploadService == null && uploadService == null) {
+            throw new SmackException("No upload service specified or discovered.");
+        }
+
+        if (uploadService == null && maxFileSize != null) {
+            if (fileSize > maxFileSize) {
+                throw new IllegalArgumentException("Requested file size " + fileSize
+                        + " is greater than max allowed size " + maxFileSize);
+            }
+        }
+
+        SlotRequest slotRequest = new SlotRequest(filename, fileSize, contentType);
+        if (uploadService != null) {
+            slotRequest.setTo(uploadService);
+        } else {
+            slotRequest.setTo(defaultUploadService);
+        }
+
+        return connection().createStanzaCollectorAndSend(slotRequest).nextResultOrThrow();
+    }
+
+    private void uploadFile(File file, URL putUrl, UploadProgressListener listener) throws IOException {
+        final HttpURLConnection urlConnection = (HttpURLConnection) putUrl.openConnection();
+        urlConnection.setRequestMethod("PUT");
+        urlConnection.setUseCaches(false);
+        urlConnection.setDoOutput(true);
+        urlConnection.setRequestProperty("Content-Type", "application/octet-stream;");
+        OutputStream outputStream = urlConnection.getOutputStream();
+
+        long bytesSend = 0;
+
+        long fileSize = file.length();
+        if (listener != null) {
+            listener.onUploadProgress(0, fileSize);
+        }
+
+        BufferedInputStream inputStream = new BufferedInputStream(new FileInputStream(file));
+
+        byte[] buffer = new byte[4096];
+        int bytesRead;
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+            outputStream.write(buffer, 0, bytesRead);
+            outputStream.flush();
```

Comment:
Wouldn't it be better to flush after the loop? A flush operation may degrade to a syscall, which is expensive.